In [11]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
pd.set_option("display.max_columns", None)


In [12]:
BASE_PATH = "/kaggle/input/uidai-hackathon"

def load_and_concat(folder_name):
    folder_path = os.path.join(BASE_PATH, folder_name,folder_name)
    
    csv_files = [
        os.path.join(folder_path, f)
        for f in os.listdir(folder_path)
        if f.endswith(".csv")
    ]
    
    df_list = [pd.read_csv(f) for f in csv_files]
    
    return pd.concat(df_list, ignore_index=True)

# Create 3 DataFrames
df_biometric = load_and_concat("api_data_aadhar_biometric")
df_demographic = load_and_concat("api_data_aadhar_demographic")
df_enrolment = load_and_concat("api_data_aadhar_enrolment")

# Quick sanity check
print("Biometric:", df_biometric.shape)
print("Demographic:", df_demographic.shape)
print("Enrolment:", df_enrolment.shape)


Biometric: (1861108, 6)
Demographic: (2071700, 6)
Enrolment: (1006029, 7)


# check first 5 columns of the dataframes to ensure they are loaded successfully.

In [13]:
df_biometric.head()

date state   district  pincode  bio_age_5_17  bio_age_17_
0  19-09-2025   Goa  North Goa   403502             0            4
1  19-09-2025   Goa  North Goa   403508             1            4
2  19-09-2025   Goa  North Goa   403513             2            0
3  19-09-2025   Goa  North Goa   403527             2            2
4  19-09-2025   Goa  South Goa   403601             7            3

In [14]:
df_demographic.head()

date        state    district  pincode  demo_age_5_17  demo_age_17_
0  19-09-2025  Maharashtra      Satara   415517              0             2
1  19-09-2025  Maharashtra      Satara   415518              0             2
2  19-09-2025  Maharashtra      Satara   415520              0             3
3  19-09-2025  Maharashtra      Satara   415539              1             5
4  19-09-2025  Maharashtra  Sindhudurg   416510              4            37

In [15]:
df_demographic.head()

date        state    district  pincode  demo_age_5_17  demo_age_17_
0  19-09-2025  Maharashtra      Satara   415517              0             2
1  19-09-2025  Maharashtra      Satara   415518              0             2
2  19-09-2025  Maharashtra      Satara   415520              0             3
3  19-09-2025  Maharashtra      Satara   415539              1             5
4  19-09-2025  Maharashtra  Sindhudurg   416510              4            37

# Are there any missing entries?

In [16]:
df_enrolment.isnull().sum()

date              0
state             0
district          0
pincode           0
age_0_5           0
age_5_17          0
age_18_greater    0
dtype: int64

In [17]:
df_demographic.isnull().sum()

date             0
state            0
district         0
pincode          0
demo_age_5_17    0
demo_age_17_     0
dtype: int64

In [18]:
df_biometric.isnull().sum()

date            0
state           0
district        0
pincode         0
bio_age_5_17    0
bio_age_17_     0
dtype: int64

What do the datasets answer?

1. `df_enrolment`: How many people applied for making Aadhar card?
2. `df_demographic`: How many people updated their Aadhar card information?
3. `df_biometric`: How many people performed Biometric for aadhar card?

# Is any column Unusual/Questionable to use?

In [33]:
df_enrolment.columns, df_demographic.columns, df_biometric.columns

(Index(['date', 'state', 'district', 'pincode', 'age_0_5', 'age_5_17',
        'age_18_greater', 'state_clean'],
       dtype='object'),
 Index(['date', 'state', 'district', 'pincode', 'demo_age_5_17', 'demo_age_17_',
        'state_clean'],
       dtype='object'),
 Index(['date', 'state', 'district', 'pincode', 'bio_age_5_17', 'bio_age_17_',
        'state_clean'],
       dtype='object'))

I assume, `age_5_17` conveys childrens' information and `age_17_/age_18_greater` convey adults' information.
The column `age_0_5` is **unusual** here. If the information for such a young age group was recorded, it must be done for all the three i.e demographic, biometric and enrolment, but it is absent in the other two.

#### _System Improvements_ : Maintaining integrity in data collection is crucial, as data is one of the most valuable assets in today’s world. All age categories should be uniformly defined and recorded across related datasets to ensure accuracy, reliability, and meaningful analysis. Inconsistencies such as missing age groups can lead to misinterpretation, reduced analytical validity, and flawed policy decisions. Standardized data collection frameworks and validation checks should be implemented to prevent such discrepancies.

# Are there any pincodes/areas in India, which are not available in other tables?

In [20]:
bio_pins = set(df_biometric["pincode"].dropna().unique())
demo_pins = set(df_demographic["pincode"].dropna().unique())
enrol_pins = set(df_enrolment["pincode"].dropna().unique())
common_pins = bio_pins & demo_pins & enrol_pins

print("Common pincodes in all three datasets:", len(common_pins))
missing_in_bio = enrol_pins - bio_pins
print("Missing in Biometric:", len(missing_in_bio))
missing_in_demo = enrol_pins - demo_pins
print("Missing in Demographic:", len(missing_in_demo))
missing_in_enrol = bio_pins - enrol_pins
print("Missing in Enrolment:", len(missing_in_enrol))


Common pincodes in all three datasets: 19432
Missing in Biometric: 23
Missing in Demographic: 16
Missing in Enrolment: 267


#### I assume, this is not a System's defect, those enrolments might have been completed earlier. For this analysis, I will skip those missing areas to ensure integrity throughout.

In [21]:
# Get the set of enrolment pincodes
enrol_pins = set(df_enrolment["pincode"])

print("Biometric rows before:", len(df_biometric))
# Filter biometric dataset to keep only rows where pincode is in enrolment pincodes
df_biometric = df_biometric[df_biometric['pincode'].isin(enrol_pins)].copy()
print("Biometric rows after :", len(df_biometric))

print("Demographic rows before:", len(df_demographic))
# Filter demographic dataset similarly
df_demographic = df_demographic[df_demographic['pincode'].isin(enrol_pins)].copy()
print("Demographic rows after :", len(df_demographic))


Biometric rows before: 1861108
Biometric rows after : 1859497
Demographic rows before: 2071700
Demographic rows after : 2069426


In [22]:
df_demographic.to_csv('demographic_data_updated.csv', index=False)
df_enrolment.to_csv('enrolment_data_updated.csv', index=False)
df_biometric.to_csv('biometric_data_updated.csv', index=False)

# Are the states clean enough?

In [23]:
df_demographic.state.unique(),df_biometric.state.unique(),df_enrolment.state.unique()

(array(['Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram', 'Nagaland',
        'ODISHA', 'Odisha', 'Orissa', 'Pondicherry', 'Puducherry',
        'Punjab', 'Rajasthan', 'Tamil Nadu', 'Telangana', 'Tripura',
        'Uttar Pradesh', 'Sikkim', 'Uttarakhand', 'West  Bengal',
        'West Bengal', 'Andaman & Nicobar Islands', 'odisha',
        'Andaman and Nicobar Islands', 'Andhra Pradesh',
        'Arunachal Pradesh', 'Assam', 'Bihar', 'Chandigarh',
        'Chhattisgarh', 'Dadra and Nagar Haveli', 'Delhi', 'Goa',
        'Gujarat', 'Haryana', 'Himachal Pradesh', 'Jammu & Kashmir',
        'Jammu and Kashmir', 'Jharkhand', 'Karnataka', 'Kerala',
        'Lakshadweep', 'Madhya Pradesh', 'Ladakh',
        'Dadra and Nagar Haveli and Daman and Diu', 'Daman and Diu',
        'west Bengal', 'Daman & Diu', 'Dadra & Nagar Haveli',
        'West Bangal', 'Westbengal', 'andhra pradesh', 'WESTBENGAL',
        'WEST BENGAL', 'West bengal', 'West Bengli', 'BALANAGAR',
        'Uttaranchal', '100000'

#### Surely not **clean**!!

In [24]:
mapping = {
    'Andhra Pradesh': [
        'andhra pradesh',
        'Andhra Pradesh',
        'Madanapalle',         

    ],

    'Jammu and Kashmir': [
        'Jammu & Kashmir',
        'Jammu and Kashmir'
    ],

    'Odisha': [
        'ODISHA',
        'Odisha',
        'Orissa',
        'odisha'
    ],

    'Pondicherry': [
        'Pondicherry',
        'Puducherry'
    ],

    'West Bengal': [
        'WEST BENGAL',
        'West Bengal',
        'west Bengal',
        'West  Bengal',
        'West Bangal',
        'Westbengal',
        'WESTBENGAL',
        'West bengal',
        'West Bengli'
    ],

    'Andaman and Nicobar Islands': [
        'Andaman & Nicobar Islands',
        'Andaman and Nicobar Islands',
      ],

    'The Dadra And Nagar Haveli And Daman And Diu': [
        'Dadra and Nagar Haveli and Daman and Diu',
        'Daman and Diu',
        'Daman & Diu',
        'Dadra & Nagar Haveli'
    ],

    'Uttarakhand': [
        'Uttaranchal'
    ],

    'Chhattisgarh': [
        'Chhatisgarh'
    ],

    'Karnataka': [
        'Puttenahalli'   
    ],

    'Maharashtra': [
        'Nagpur'        
    ],

    'Bihar': [
        'Darbhanga'       
    ]
,
    'Telangana': [
        'BALANAGAR'
    ],
    'Rajasthan': [
        'Jaipur'
    ]
,
    'Tamil Nadu': [
        'Tamilnadu',
        'Raja Annamalai Puram' 
    ],
    'The Dadra And Nagar Haveli And Daman And Diu': [
        'Dadra and Nagar Haveli and Daman and Diu',
        'Daman and Diu',
        'Daman & Diu',
        'Dadra & Nagar Haveli',
        'Dadra and Nagar Haveli'
    ],
}


In [25]:
STATE_LOOKUP = {
    variant.lower(): standard
    for standard, variants in mapping.items()
    for variant in variants
}
def normalize_state(value):
    if not isinstance(value, str):
        return value

    value_clean = value.strip().lower()
    return STATE_LOOKUP.get(value_clean, value)


In [26]:
df_biometric['state_clean'] = df_biometric['state'].apply(normalize_state)
df_demographic['state_clean'] = df_demographic['state'].apply(normalize_state)
df_enrolment['state_clean'] = df_enrolment['state'].apply(normalize_state)

In [27]:
df_biometric.state_clean.unique()

array(['Goa', 'Gujarat', 'Haryana', 'Himachal Pradesh',
       'Jammu and Kashmir', 'Jharkhand', 'Karnataka', 'Kerala',
       'Andaman and Nicobar Islands', 'Andhra Pradesh', 'Mizoram',
       'Nagaland', 'Odisha', 'Pondicherry', 'Punjab', 'Rajasthan',
       'Sikkim', 'Tamil Nadu', 'Telangana', 'Tripura', 'Uttar Pradesh',
       'Uttarakhand', 'West Bengal', 'Ladakh', 'Lakshadweep',
       'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya',
       'Arunachal Pradesh', 'Assam', 'Bihar', 'Chandigarh',
       'Chhattisgarh', 'The Dadra And Nagar Haveli And Daman And Diu',
       'Delhi'], dtype=object)

In [28]:
df_enrolment.state_clean.unique()

array(['Meghalaya', 'Karnataka', 'Uttar Pradesh', 'Bihar', 'Maharashtra',
       'Haryana', 'Rajasthan', 'Punjab', 'Delhi', 'Madhya Pradesh',
       'West Bengal', 'Assam', 'Uttarakhand', 'Gujarat', 'Andhra Pradesh',
       'Tamil Nadu', 'Chhattisgarh', 'Jharkhand', 'Nagaland', 'Manipur',
       'Telangana', 'Tripura', 'Mizoram', 'Jammu and Kashmir',
       'Chandigarh', 'Sikkim', 'Odisha', 'Kerala',
       'The Dadra And Nagar Haveli And Daman And Diu',
       'Arunachal Pradesh', 'Himachal Pradesh', 'Goa', 'Ladakh',
       'Andaman and Nicobar Islands', 'Pondicherry', 'Lakshadweep',
       '100000'], dtype=object)

In [29]:
df_demographic.state_clean.unique()

array(['Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram', 'Nagaland',
       'Odisha', 'Pondicherry', 'Punjab', 'Rajasthan', 'Tamil Nadu',
       'Telangana', 'Tripura', 'Uttar Pradesh', 'Sikkim', 'Uttarakhand',
       'West Bengal', 'Andaman and Nicobar Islands', 'Andhra Pradesh',
       'Arunachal Pradesh', 'Assam', 'Bihar', 'Chandigarh',
       'Chhattisgarh', 'The Dadra And Nagar Haveli And Daman And Diu',
       'Delhi', 'Goa', 'Gujarat', 'Haryana', 'Himachal Pradesh',
       'Jammu and Kashmir', 'Jharkhand', 'Karnataka', 'Kerala',
       'Lakshadweep', 'Madhya Pradesh', 'Ladakh', '100000'], dtype=object)

#### Almost cleaned now!
#### _System Improvements_: Maintain a dropdown for states. Some values were not even states, like Jaipur, BALANAGAR, etc. 
#### **Anamoly detected** :The pincode,"100000", was given as information for all the three columns, i.e. state, district and pincode, which is clearly not following the standard meaning of the columns. Additionally, There is no confirmation whether it is a valid entry or a placeholder. Certainly, one possibility that exists, convey that someone could be misusing the power of Aadhar Card from this pincode. Hence, this is a much needed system improvement, that data integrity is maintained in the recorded data.

In [30]:
df_enrolment[df_enrolment['state']=='100000']

date   state district  pincode  age_0_5  age_5_17  \
23108   02-09-2025  100000   100000   100000        0         0   
46946   03-09-2025  100000   100000   100000        0         0   
97816   08-09-2025  100000   100000   100000        0         0   
115798  09-09-2025  100000   100000   100000        0         0   
153156  11-09-2025  100000   100000   100000        0         0   
160195  12-09-2025  100000   100000   100000        0         0   
261778  19-09-2025  100000   100000   100000        0         0   
272731  20-09-2025  100000   100000   100000        0         0   
470934  24-10-2025  100000   100000   100000        0         1   
762744  15-11-2025  100000   100000   100000        0         0   
780936  16-11-2025  100000   100000   100000        0         0   
791621  17-11-2025  100000   100000   100000        0         0   
823985  19-11-2025  100000   100000   100000        0         0   
861631  15-12-2025  100000   100000   100000        0         0   
914457  24-12-2025  100000   100000   100000        0         0   
929221  25-12-2025  100000   100000   100000        0         0   
947831  26-12-2025  100000   100000   100000        0         0   
959114  27-12-2025  100000   100000   100000        0         0   
960973  27-12-2025  100000   100000   100000        0         0   
968919  28-12-2025  100000   100000   100000        0         0   
976509  29-12-2025  100000   100000   100000        0         0   
988283  31-12-2025  100000   100000   100000        0         0   

        age_18_greater state_clean  
23108                3      100000  
46946                1      100000  
97816                1      100000  
115798               1      100000  
153156               2      100000  
160195               2      100000  
261778               1      100000  
272731               1      100000  
470934               0      100000  
762744               3      100000  
780936               2      100000  
791621               4      100000  
823985               2      100000  
861631             161      100000  
914457               1      100000  
929221               4      100000  
947831               3      100000  
959114               4      100000  
960973               4      100000  
968919               5      100000  
976509               8      100000  
988283               4      100000

In [31]:
df_biometric[df_biometric['state']=='100000']

Empty DataFrame
Columns: [date, state, district, pincode, bio_age_5_17, bio_age_17_, state_clean]
Index: []

In [32]:
df_demographic[df_demographic['state']=='100000']

date   state district  pincode  demo_age_5_17  demo_age_17_  \
295161   23-12-2025  100000   100000   100000              0             1   
1507370  20-12-2025  100000   100000   100000              0             1   

        state_clean  
295161       100000  
1507370      100000

#### Look at the spike on 15th Dec, 2025. **161** enrolments is not a usual value. The Goverment must look into it.

# UIDAI Hackathon Data Analysis Report

## Executive Summary

Analysis of three critical UIDAI datasets: Biometric Authentication, Demographic Updates, and Enrolment Records spanning September-December 2025.

---

## Key Insights and Findings

### 1. Data Integrity Issues

#### **Age Group Inconsistency**
- **Issue**: `age_0_5` column exists only in enrolment dataset
- **Impact**: Inconsistent age categorization across related datasets
- **Observation**: If data for age 0-5 is collected, it should be uniformly recorded across all three datasets (biometric, demographic, enrolment)

#### **Geographic Coverage Gaps**
- Common pincodes across all datasets: **19,432**
- Missing in Biometric: **23 pincodes**
- Missing in Demographic: **16 pincodes**
- Missing in Enrolment: **267 pincodes**
- **Analysis**: Likely represents completed enrolments from earlier periods

### 2. Critical Data Quality Problems

#### **State Name Standardization Crisis**
Found **multiple variations** of state names including:
- West Bengal: 9 different spellings (WEST BENGAL, Westbengal, West Bangal, etc.)
- Odisha: 4 variations (ODISHA, Odisha, Orissa, odisha)
- Invalid entries: City names as states (Jaipur, Nagpur, BALANAGAR, Madanapalle, etc.)

#### **Anomalous Entry: Pincode "100000"**
- Used as value for state, district, AND pincode columns
- **22 enrolment records** with this anomaly
- **2 demographic records** affected
- Appears to be a placeholder pincode
- **Security Concern**: Potential misuse of Aadhar Card or data entry error

### 3. Suspicious Activity Detected

#### **Unusual Spike on December 15, 2025**
- **161 enrolments** recorded for pincode "100000" in a single day
- Dramatic deviation from typical pattern (1-8 enrolments/day)
- **Requires immediate investigation** for potential fraud or system manipulation

---

---

## Dataset Statistics

| Dataset | Total Records | Cleaned Records | Columns |
|---------|--------------|-----------------|---------|
| **Enrolment** | 1,006,029 | 1,006,029 | 7 |
| **Demographic** | 2,071,700 | 2,069,426 | 6 |
| **Biometric** | 1,861,108 | 1,859,497 | 6 |

---

## Impact Statement

**Data is one of the most valuable assets in today's world.** Maintaining integrity in data collection is crucial for:
- Accurate policy decisions
- Reliable analytical insights
- Public trust in UIDAI systems
- Prevention of identity fraud
- Effective resource allocation

Poor data quality leads to:
- ❌ Misinterpretation and flawed analysis
- ❌ Reduced analytical validity
- ❌ Incorrect government policy decisions
- ❌ Potential security vulnerabilities

---

## Conclusion

While the UIDAI system processes millions of records successfully, the identified data quality issues present significant risks. Implementing the recommended standardization and validation controls will:

- Ensure data accuracy and reliability
- Prevent fraudulent activities
- Enable better decision-making
- Maintain public trust in the Aadhaar system

**Immediate action required** on the December 15th anomaly and pincode "100000" investigation.

---

*Analysis conducted for UIDAI Hackathon | Data Period: March-December 2025*